### PR TITLE
networkd: add new KeepConfiguration= setting

### LINKED ISSUE #12050: systemd-networkd prunes floating virtual IPs for high-availability environments
**Is your feature request related to a problem? Please describe.**

When maintaining virtual floating IPs (VIPs) on high-available systems, I'm seeing them pruned when restarting `systemd-networkd` (even with no config changes) which results in a downtime.

This problem is especially severe when DHCP is used, as it is also triggered when the DHCP lease gets renewed (some cloud providers use DHCP by default to configure network interfaces). But also on static IP configurations, systemd-networkd might get restarted from time to time (automatic security patches, etc.).

I'm aware that this is a [feature](https://github.com/systemd/systemd/issues/780#issuecomment-126067240) (which I agree is useful for desktops, etc.). In high-available environments with VIPs this backfires when using Linux distributions that rely on `systemd-networkd` (like Ubuntu, via [netplan](netplan.io)).

**Describe the solution you'd like**

I'm suggesting a flag that disables this behaviour, or would allow whitelisting certain VIPs to prevent them from purging. I've tried `CriticalConnection=true`, but this doesn't prevent the pruning.

**Describe alternatives you've considered**

1. Using a `dummy` interface and rely on weak host mode (ES): This doesn't work, as the gratitious ARPs/ unsolicited neighbour adverts can't be send using a dummy interface. There's another issue with this approach, which I described in https://github.com/systemd/systemd/issues/12051
2. Filing a patch to keepalived to support sending out NA/GARP via another interface (which is currently considered by them, see https://github.com/acassen/keepalived/issues/1170, but might not work
3. Migrating back to `ifupdown`
4. Migrating to a distribution without `systemd-networkd`
5. Migrating to BSD

### PR DESCRIPTION
**Pull request by ssahani**
This looks pretty legit. networkd drops the foreign addresses upon
restart. This becomes a problem in cases where the additional address
is used as a virtual or floating IP in higih-available environments
(where a master IP is swapped to a different service upon failovers,
e.g. by services like Pacemaker, keepalived and alike, usually using
protocols like Corosync, VRRP or CARP). Those addresses are not handled
by systemd-networkd and are therefore purged upon DHCP renewals or
when an update triggers a restart of systemd-networkd, therefore
potentially resulting in downtimes or at least unecessary failovers.

closes #12050

### CODE REVIEW

**Comment by yuwata** [comment]
I think this also fixes #11575.

**Comment by yuwata** [comment]
So, the setting is useful not only for HA environment. How about moving `CriticalConnection=` from `[DHCP]` section to `[Network]` section, and introduce `CriticalConnection=static` (though the name `static` may be misleading, but I have no idea)?
- `CriticalConnection=yes`:  do not drop configs on start, and DHCP renew.
- `CriticalConnection=static`: do not drop configs on start, but DHCP addresses are dropped on renew.
- `CriticalConnection=no`: drop configs on start.

**Comment by ssahani** [comment]
We can't move ```CriticalConnection=``` to [Network] section because of backward compatibility. We actually do not allow any other servers to configure same interface with networkd because they may cause lots of misconfiguration. But this case makes a tons of sense and describes the purpose ```HA```. We might want to rename but not sure.

**Comment by yuwata** [comment]
> We can't move `CriticalConnection=` to [Network] section because of backward compatibility. 

Of course, `CriticalConnection=` in `[DHCP]` still needs to be supported.

> We actually do not allow any other servers to configure same interface with networkd because they may cause lots of misconfiguration. But this case makes a tons of sense and describes the purpose `HA`. We might want to rename but not sure.

I think the name `CriticalConnection=` already suggests it is something special. Of course, the man page needs to warn about that.

Anyway, apart from the name of the setting, I like this change.

@johannbg You do not need to wait the patch to be backported. Setting `CriticalConnection=yes` in `[DHCP]` section (even if you do not use DHCP) should work.

**Comment by yuwata** [comment]
One reason I do not like the name `HighAvailability=` is that the setting does not provide HA, but just _for_ HA systems.

**Comment by ssahani** [comment]
```AllowHighAvailability=``` ?

**Comment by johannbg** [comment]
@yuwata  I'm in midst of setting up test ipv4/ipv6 environment on F30 networkd only setup to test the classic HAProxy + Keepalived setup so I'll try it out but this will need to be backported regardless.

@ssahani Presumably in the future HA application stacks would want to have networkd manage the network settings ( and associated sysctl settings ) for them?  ( I would not be surprised if there is very little things missing from networkd to do that )

If so it would need to be something that goes nicely with the rest of the [Network] section so something like below instead of HighAvailability= ( I personally think HighAvailability= is quite self explanatory thou )
 
Redundant=
Primary=/Secondary=
VIP=
FIP=
VirtualIP=
FloatingIP=
VirtualAddresss=
FloatingAddress=

**Comment by ssahani** [comment]
> If so it would need to be something that goes nicely with the rest of the [Network] section so something like below instead of HighAvailability= ( I personally think HighAvailability= is quite self explanatory thou )
Redundant=
Primary=/Secondary=
VIP=
FIP=
VirtualIP=
FloatingIP=
VirtualAddresss=
FloatingAddress=

Let's just make it generic and something comes in future also we need not add / remove stuff. Other application which have same requirements and we can't just keep on adding them no ?

**Comment by johannbg** [comment]
Redundant=True is probably the most generic terminology which would cover HA amongst other things. 
networkd probably requires an capability-based IPC support (dbus/kdbus/bus1 whatever ) before applications and application stacks can make full use of what networkd is capable of and make such future become a reality.

**Comment by ssahani** [comment]
>Redundant=True is probably the most generic terminology which would cover HA amongst other things.

Well I am not sure @yuwata WDUT?

>networkd probably requires an capability-based IPC support (dbus/kdbus/bus1 whatever ) before applications and application stacks can make full use of what networkd is capable of and make such future become a reality.

yeah varlink just went in and we have dbus stuff to do.

**Comment by chr4** [comment]
In case it helps: 

In my tests (although I'm using [netplan](https://netplan.io)), setting `critical: true` (which sets `CriticalConnection=True` according to the [source](https://git.launchpad.net/netplan/tree/src/networkd.c#n547)) did not prevent purging the addresses upon a `systemctl restart systemd-networkd`.

```yaml
network:
    version: 2
    ethernets:
        ens3:
            dhcp4: true
            critical: true
```

```bash
ip a a 1.2.3.4/32 dev eth0
systemctl restart systemd-networkd
ip a |grep 1.2.3.4
```

(Sorry, had no time to reproduce this using plain systemd-networkd.)

**Comment by keszybz** [comment]
> So, the setting is useful not only for HA environment.

Yes, this is not tied in any particular way to HA. The name should be generic.

> How about moving CriticalConnection= from [DHCP] section to [Network] section, and introduce CriticalConnection=static

I think so too. Let's go one step futher and use a name that actually describes in plain words what this option does:
`KeepConfiguration=yes|no|dhcp|static...`,
where "yes" is the same as "dhcp, static" and `KeepConfiguration=dhcp` is the same as `CriticalConnection=yes`.

**Comment by yuwata** [comment]
@ssahani Are you working on this? If not, I will try to implement the option @keszybz suggests.

**Comment by ssahani** [comment]
yes I would like to work it over the weekend

**Comment by ssahani** [comment]
I have updated PTAL @yuwata @keszybz .

**Comment by yuwata** [comment]
Hmm. How about the following?
- `KeepConfiguration=dhcp-on-stop` : do not stop clients on stopping `networkd`. Defaults to this option for backward compatibility.
- `KeepConfiguration=dhcp` : equivalent to `CriticalConnection=yes`. Including `dhcp-on-stop`.
- `KeepConfiguration=static` : do not remove addresses and routes on starting up.
- `KeepConfiguration=yes` : Including the above three options.

And then, parser of `CriticalConnection=` should be updated as an alias of `KeepConfiguration=dhcp`.

@ssahani If you are busy recently, then can I update your commit accordingly?

BTW, should we split `dhcp` into `dhcp4` and `dhcp6`? WDYT?

**Comment by yuwata** [comment]
@ssahani and @keszybz I've slightly updated @ssahani's commit and added several follow-ups. PTAL.

@ssahani Your original commit is saved at [here](https://github.com/yuwata/systemd/tree/ssahani-high-availability-12050-backup).

File: src/network/networkd-network.c [hunk 1]
```diff
@@ -239,6 +239,22 @@ int network_verify(Network *network) {
...
+                        network->keep_configuration = KEEP_CONFIGURATION_YES;
+                else
+                        /* For backward compatibility, we do not release DHCP addresses on stop. */
+                        network->keep_configuration = KEEP_CONFIGURATION_DHCP_ON_STOP;
+        }
+
+        if (network->keep_configuration < 0)
+                /* For backward compatibility, we do not release DHCP addresses on stop. */
+                network->keep_configuration = KEEP_CONFIGURATION_DHCP_ON_STOP;
+
```
Comment (ssahani) [inline src/network/networkd-network.c:257]: This is seems wrong. It's not backward compatibility. It's was a bug. As soon as DHCP client stops those addresses acquired by client no more valid and hence should be dropped.

**Comment by ssahani** [comment]
Thanks for working on this.

> * `KeepConfiguration=dhcp-on-stop` : do not stop clients on stopping `networkd`. Defaults to this for backward compatibility.

It's wrong . As soon as DHCP client stops those addresses are not valid any more. I have already explained it.

> * `KeepConfiguration=dhcp` : equivalent to `CriticalConnection=yes`. Including `dhcp-on-stop`.

Yes .
> * `KeepConfiguration=static` : do not remove addresses and routes on starting up.
> * `KeepConfiguration=yes` : Including the above three options.

Yes rest are OK

**Comment by yuwata** [comment]
@ssahani Thank you for the comments.

> > * `KeepConfiguration=dhcp-on-stop` : do not stop clients on stopping `networkd`. Defaults to this for backward compatibility.
> 
> It's wrong . As soon as DHCP client stops those addresses are not valid any more. I have already explained it.

Theoretically yes. However,

- If we do not provide the option and 'fix' the bug, then many network dependent systems have to set `KeepConfiguration=dhcp` (or `CriticalConnection=yes`) when people needs to restart `networkd`. But setting `KeepConfiguration=dhcp` may introduce unwanted behavior. Or, in other words, we cannot safely restart `networkd` without the option.

- The 'bug' causes real buggy situation almost only when `networkd` just stopped, not restarted. Then, the lease address may be re-assigned to another machine. However, I hope, the usual usecase of `networkd` is long running daemon, that is, the daemon is only restarted. not stopped so long time. So, the 'bug' merely causes problem.

- Moreover, the 'bug' has existed for long time... So, I think we should keep the theoretical 'bug' by default.

**Comment by ssahani** [comment]
> Theoretically yes. However,
>

Not sure what is theoretically mean. It a BUG. 
 
> * If we do not provide the option and 'fix' the bug, then many network dependent systems have to set `KeepConfiguration=dhcp` (or `CriticalConnection=yes`) when people needs to restart `networkd`. But setting `KeepConfiguration=dhcp` may introduce unwanted behavior. Or, in other words, we cannot safely restart `networkd` without the option.

It a right behaviour. Once networked stops it should drop the DHCP addresses.

> * The 'bug' causes real buggy situation almost only when `networkd` just stopped, not restarted. Then, the lease address may be re-assigned to another machine. However, I hope, the usual usecase of `networkd` is long running daemon, that is, the daemon is only restarted. not stopped so long time. So, the 'bug' merely causes problem.

How we can ensure that. This is a complete myth.

> * Moreover, the 'bug' has existed for long time... So, I think we should keep the theoretical 'bug' by default.

No It's a practical BUG.

**Comment by yuwata** [comment]
Now I understand that 'stopping DHCP client' != 'stopping networkd'. And I guess one reason why our discussion could not converge is I did not understand that. Sorry.

@ssahani Could you take a look once more? Now, `KEEP_CONFIGURATION_DHCP_ON_STOP` option is used only when stopping `networkd`, hopefully restarting `networkd`.

Of course, the current version did not answer your request... But, I hope it is better than the previous one.

**Comment by ssahani** [comment]
Yes precisely dynamic addresses are valid when we keep talking with the server. When the communication breaks we can not ensure those addresses are valid. Thanks this is better for sure. LGTM.

**Comment by yuwata** [comment]
Yey! All green now. @ssahani Thank you for the long discussion. @keszybz Could you take a final look?

**Comment by keszybz** [comment]
I think this can go in. It is nice functionality that can be useful in various cases.

Nevertheless, I think that we should rework how DHCP addresses are handled. In particular, when restarting, we shouldn't drop any addresses before the restart. All state is serialized, so it's just a question of adjusting to new configuration after restart. I don't want users to set KeepConfiguration= just be able to restart sytemd-networkd painlessly.

I'll try to do the rebase now myself and force-push to your branch to redo the CI.

**Comment by yuwata** [comment]
Rebased.

**Pull request merged**